File: web-ui-with-angular-cli/src/app/patient-manager/patient-visit/models/eo-visit.ts
import { InsuranceMaster } from '../../patients/models/insurance-master';
import { Case } from '../../cases/models/case';
import { IEventWrapper } from '../../../commons/models/i-event-wrapper';
import { ScheduledEventInstance } from '../../../commons/models/scheduled-event-instance';
import { ScheduledEvent } from '../../../commons/models/scheduled-event';
import { Record } from 'immutable';
import * as moment from 'moment';
import * as _ from 'underscore';
import * as RRule from 'rrule';
import { VisitStatus } from './enums/visit-status';
import { Patient } from '../../../patient-manager/patients/models/patient';
import { Doctor } from '../../../medical-provider/users/models/doctor';

const EoVisitRecord = Record({
    id: 0,
    calendarEventId: 0,
    location: null,
    locationId: 0,
    isEoVisitType: true,
    doctor: null,
    patient: null,
    doctorId: 0,
    patientId: null,
    caseId: null,
    insuranceMaster: null,
    insuranceProviderId: 0,
    VisitCreatedByCompanyId: null,
    eventStart: null,
    eventEnd: null,
    notes: '',
    visitStatusId: VisitStatus.SCHEDULED,
    calendarEvent: null,
    isDeleted: false,
    createByUserId: 0,
    updateByUserId: 0,
    createDate: null, //Moment
    updateDate: null,//Moment
    visitTimeStatus: false,
    visitUpdateStatus: false
    
});


export class EoVisit extends EoVisitRecord {

    id: number;
    calendarEventId: number;
    location: Location;
    locationId: number;
    isEoVisitType: boolean;
    doctor: Doctor;
    patient: Patient;
    doctorId: number;
    patientId: number;
    caseId: number;
    insuranceMaster: InsuranceMaster;
    insuranceProviderId: number;
    VisitCreatedByCompanyId: number;
    eventStart: moment.Moment;
    eventEnd: moment.Moment;
    notes: string;
    visitStatusId: VisitStatus;
    calendarEvent: ScheduledEvent;
    isDeleted: boolean;
    createByUserId: number;
    updateByUserId: number;
    createDate: moment.Moment;
    updateDate: moment.Moment;
    visitTimeStatus: boolean;
    visitUpdateStatus: boolean;

    constructor(props) {
        super(props);
    }

    get isOriginalVisit(): boolean {
        return !this.eventStart ? true : false;
    }

    get isExistingVisit(): boolean {
        return !this.isOriginalVisit;
    }

    get eventColor(): string {
        return '#7A3DB8';
    }

    get visitDisplayString(): string {
        let visitInfo: string = ``;

        if (this.doctorId && this.doctor) {
            visitInfo = `${visitInfo}Doctor Name: ${this.doctor.user.firstName} ${this.doctor.user.lastName}`;
        }

        return visitInfo;
    }

    get visitDisplayStringForDoctor(): string {
        let visitInfo: string = ``;

        if (this.insuranceProviderId && this.insuranceMaster) {
            visitInfo = `${visitInfo}Insurance Name: ${this.insuranceMaster.companyName}`;
        }

        return visitInfo;
    }

    get visitStatusLabel(): string {
        return EoVisit.getvisitStatusLabel(this.visitStatusId);
    }

    static getvisitStatusLabel(visitStatus: VisitStatus): string {
        switch (visitStatus) {
            case VisitStatus.SCHEDULED:
                return 'Scheduled';
            case VisitStatus.COMPLETE:
                return 'Complete';
            case VisitStatus.RESCHEDULE:
                return 'Rescheduled';
            case VisitStatus.NOSHOW:
                return 'Noshow';
        }
    }
}

File: Midas/MIDAS.GBX/GBBusinessObjects/EOVisit/EOVisit.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MIDAS.GBX.BusinessObjects
{
    public class EOVisit : GbObject
    {
        [JsonProperty("doctorId")]
        public int? DoctorId { get; set; }

        [JsonProperty("patientId")]
        public int? PatientId { get; set; }

        [JsonProperty("visitCreatedByCompanyId")]
        public int? VisitCreatedByCompanyId { get; set; }

        [JsonProperty("insuranceProviderId")]
        public int? InsuranceProviderId { get; set; }

        [JsonProperty("calendarEventId")]
        public int? CalendarEventId { get; set; }

        [JsonProperty("visitStatusId")]
        public int? VisitStatusId { get; set; }

        [JsonProperty("eventStart")]
        public DateTime? EventStart { get; set; }

        [JsonProperty("eventEnd")]
        public DateTime? EventEnd { get; set; }

        [JsonProperty("notes")]
        public string Notes { get; set; }

        [JsonProperty("calendarEvent")]
        public CalendarEvent CalendarEvent { get; set; }

        [JsonProperty("doctor")]
        public Doctor Doctor { get; set; }        

        [JsonProperty("medicalProvider")]
        public Company Company { get; set; }

        [JsonProperty("insuranceMaster")]
        public InsuranceMaster InsuranceMaster { get; set; }

    }
}
